### PR TITLE
null backend: fix+test resource usage after cmdbuf transition

### PR DESCRIPTION
**Pull request by kainino0x**

### CODE REVIEW

**Comment by kainino0x** [comment]
This is an issue I've been thinking of and forgetting about for months now.

This is just a test for now, because I want to discuss what the actual behavior should be.

**Comment by kainino0x** [comment]
@Kangz I added some code in the null backend which should allow this test to pass. But instead, I'm getting a segfault. Something I don't understand is happening in `CommandAllocator`: `currentBlock` and `currentPtr` are getting garbage values on one iteration of NextCommandId. Not sure how this happens; I don't see anywhere `currentBlock` could get assigned garbage, and `-fsanitizer=address -fstack-protector-all` isn't finding anything before the crash.

File: src/backend/null/NullBackend.cpp [hunk 1]
```diff
@@ -155,6 +157,89 @@ namespace null {
...
+
+    CommandBuffer::~CommandBuffer() {
+        FreeCommands(&commands);
+    }
+
+    void CommandBuffer::Execute() {
+        Command type;
+        while (commands.NextCommandId(&type)) {
+            switch (type) {
+                case Command::AdvanceSubpass:
```
Comment (Kangz) [inline src/backend/null/NullBackend.cpp]: nit: use the new and handy SkipCommands for everything except the transitions

File: src/backend/null/NullBackend.cpp [hunk 2]
```diff
@@ -155,6 +157,89 @@ namespace null {
     void Buffer::TransitionUsageImpl(nxt::BufferUsageBit currentUsage, nxt::BufferUsageBit targetUsage) {
     }
 
+    // CommandBuffer
+
+    CommandBuffer::CommandBuffer(CommandBufferBuilder* builder)
+        : CommandBufferBase(builder), device(ToBackend(builder->GetDevice())),
```
Comment (Kangz) [inline src/backend/null/NullBackend.cpp]: nit: the private device member isn't used.

File: src/tests/unittests/validation/BufferValidationTests.cpp [hunk 3]
```diff
@@ -317,3 +317,29 @@ TEST_F(BufferValidationTest, SetSubDataWrongUsage) {
     uint32_t foo = 0;
     ASSERT_DEVICE_ERROR(buf.SetSubData(0, 1, &foo));
 }
+
+// Test that command buffer submit changes buffer usage
```
Comment (Kangz) [inline src/tests/unittests/validation/BufferValidationTests.cpp]: nit: Given all the tiny details of usage validation, I suggest we make a UsageValidationTests file.

File: src/backend/null/NullBackend.cpp (hunk 2 above)
Comment (kainino0x) [inline src/backend/null/NullBackend.cpp]: Done

File: src/tests/unittests/validation/BufferValidationTests.cpp (hunk 3 above)
Comment (kainino0x) [inline src/tests/unittests/validation/BufferValidationTests.cpp]: Done

File: src/backend/null/NullBackend.cpp (hunk 1 above)
Comment (kainino0x) [inline src/backend/null/NullBackend.cpp]: Done

**Pull request merged**